File: src/models/map.ts
import { ddnet, master, dbQuery } from "../lib/database.js"
import { handleErrors, groupBy, splitMappers } from "../lib/misc.js"
import getLogger from "../lib/logger.js"
import colors from "colors"

const log = getLogger("Database  |", colors.yellow)

const Map = {
    categories: [
        "Any",
        "Novice",
        "Moderate",
        "Brutal",
        "Insane",
        "Dummy",
        "DDmaX.Easy",
        "DDmaX.Next",
        "DDmaX.Pro",
        "DDmaX.Nut",
        "Oldschool",
        "Solo",
        "Race",
        "Fun"
    ],
    /**
     * Get information about a map
     * @param {string} map
     * @param {integer} top - How many ranks to return?
     * @returns {Promise<Array>}
     */
    info: handleErrors(async (map: string) => {
        let info = await dbQuery(ddnet, `
            SELECT maps.*, ROUND(mediantime) as mediantime, finishes, finishesrank FROM maps JOIN mapstats ON maps.map = mapstats.map WHERE maps.map = ?;
        `, [map], true)
        if(info)
            info.Mapper = info.Mapper.split(/, | & /)
        return info
    }, log),
    /**
     * Get rankings of a map.
     * @param {string} map
     * @param {integer} top - How many ranks to return?
     * @returns {Promise<Array>}
     */
    rankings: handleErrors(async (map: string, top: number) => {
        return await dbQuery(ddnet, `
            SELECT * FROM rankings WHERE map = ? AND rank <= ? ORDER BY rank LIMIT ?;
        `, [map, top, top])
    }, log),
    /**
     * Get teamrankings of a map.
     * @param {string} map
     * @param {integer} top - How many ranks to return?
     * @returns {Promise<Array>}
     */
    teamrankings: handleErrors(async (map: string, top: number) => {
        return groupBy(await dbQuery(ddnet, `
            SELECT * FROM teamrankings WHERE map = ? AND rank <= ? ORDER BY rank;
        `, [map, top]), "ID")
    }, log),
    /**
     * Get timecps of a map.
     * @param {string} map
     * @param {integer} top - How many ranks to return?
     * @returns {Promise<Array>}
     */
    timecps: handleErrors(async (map: string, limit: number) => {
        return await dbQuery(ddnet, `
            SELECT 
                Name, 
                cp1, cp2, cp3, cp4, cp5, cp6, cp7, cp8, cp9, cp10, cp11, cp12, cp13, cp14, cp15, cp16, cp17, cp18, cp19, cp20, cp21, cp22, cp23, cp24, cp25, 
                time 
            FROM race WHERE map = ? GROUP BY name ORDER BY time ASC LIMIT ${limit};
        `, [map])
    }, log),
    /**
     * Search for player.
     * @param {object}  search
     * @param {string}  [search.map]
     * @param {string}  [search.mapper]
     * @param {integer}  [search.limit]
     * @returns {Promise<Array>}
     */
    search: handleErrors(async (search: { mapper?: string; limit?: number; map: string }) => {
        search.mapper = search.mapper ?? "*"
        let maps = await dbQuery(ddnet, `
            SELECT * FROM maps WHERE map LIKE FORMAT('%%%s%', ?) AND (mapper GLOB FORMAT('%s', ?)
                OR mapper GLOB FORMAT('%s *', ?) 
                OR mapper GLOB FORMAT('* %s', ?)
                OR mapper GLOB FORMAT('%s, *', ?)
                OR mapper GLOB FORMAT('*, %s & *', ?)
                OR mapper GLOB FORMAT('*, %s, *', ?)) ${search.limit ? `LIMIT ${search.limit}` : ""}
        `, [search.map, search.mapper, search.mapper, search.mapper, search.mapper, search.mapper, search.mapper])

        for (const map in maps) {
            maps[map].Mapper = splitMappers(maps[map].Mapper)
        }
        
        return maps

    }, log),
}

export default Map

File: src/models/map.rs
use serde::{Deserialize, Serialize};
use sqlx::{self, types::chrono::NaiveDateTime, Pool, Postgres};
use std::fmt::Debug;

#[derive(Debug, Clone, Deserialize, Serialize, sqlx::Type)]
pub struct Map {
    pub map: String,
    pub server: String,
    pub points: i32,
    pub stars: i32,
    pub mapper: String,
    pub timestamp: Option<NaiveDateTime>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Info {
    pub map: Map,
    pub finishes: Option<i32>,
    pub finishes_rank: Option<i32>,
    pub median_time: Option<f64>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Rankings {
    pub rank: i32,
    pub timestamp: Option<NaiveDateTime>,
    pub name: String,
    pub time: f64,
    pub map: String,
    pub server: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TeamRankings {
    pub rank: i32,
    pub timestamp: Option<NaiveDateTime>,
    pub id: Vec<u8>,
    pub players: Vec<String>,
    pub time: f64,
    pub map: String,
    pub server: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TimeCps {
    pub name: String,
    pub cp1: f64,
    pub cp2: f64,
    pub cp3: f64,
    pub cp4: f64,
    pub cp5: f64,
    pub cp6: f64,
    pub cp7: f64,
    pub cp8: f64,
    pub cp9: f64,
    pub cp10: f64,
    pub cp11: f64,
    pub cp12: f64,
    pub cp13: f64,
    pub cp14: f64,
    pub cp15: f64,
    pub cp16: f64,
    pub cp17: f64,
    pub cp18: f64,
    pub cp19: f64,
    pub cp20: f64,
    pub cp21: f64,
    pub cp22: f64,
    pub cp23: f64,
    pub cp24: f64,
    pub cp25: f64,
    pub time: f64,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Playtime {
    pub name: String,
    pub seconds_played: i64,
}

impl Map {
    /// Search for a Map
    pub async fn search(
        db: &Pool<Postgres>,
        query: &str,
        n: Option<i64>,
    ) -> Result<Vec<Info>, sqlx::Error> {
        sqlx::query_file_as!(Info, "sql/map/search.sql", query, n)
            .fetch_all(db)
            .await
    }

    /// Get information about a map
    pub async fn info(db: &Pool<Postgres>, map: &str) -> Result<Info, sqlx::Error> {
        sqlx::query_file_as!(Info, "sql/map/info.sql", map)
            .fetch_one(db)
            .await
    }

    /// Get information about all maps
    pub async fn get_all(db: &Pool<Postgres>) -> Result<Vec<Map>, sqlx::Error> {
        sqlx::query_file_as!(Map, "sql/map/get_all.sql")
            .fetch_all(db)
            .await
    }

    /// Get rankings of a map
    pub async fn rankings(
        db: &Pool<Postgres>,
        map: &str,
        n: Option<i32>,
    ) -> Result<Vec<Rankings>, sqlx::Error> {
        sqlx::query_file_as!(Rankings, "sql/map/rankings.sql", map, n)
            .fetch_all(db)
            .await
    }

    /// Get teamrankings of a map
    pub async fn team_rankings(
        db: &Pool<Postgres>,
        map: &str,
        n: Option<i32>,
    ) -> Result<Vec<TeamRankings>, sqlx::Error> {
        sqlx::query_file_as!(TeamRankings, "sql/map/team_rankings.sql", map, n)
            .fetch_all(db)
            .await
    }

    /// Get timecps of a map
    pub async fn time_cps(
        db: &Pool<Postgres>,
        map: &str,
        sort_by: Option<String>,
        n: Option<i64>,
    ) -> Result<Vec<TimeCps>, sqlx::Error> {
        if sort_by.is_none() {
            sqlx::query_file_as!(TimeCps, "sql/map/time_cps.sql", map, n)
                .fetch_all(db)
                .await
        } else {
            sqlx::query_file_as!(TimeCps, "sql/map/time_cps_sort.sql", map, sort_by, n)
                .fetch_all(db)
                .await
        }
    }

    /// Get playtime of a map
    pub async fn playtime(
        db: &Pool<Postgres>,
        map: &str,
        n: Option<i64>,
    ) -> Result<Vec<Playtime>, sqlx::Error> {
        sqlx::query_file_as!(Playtime, "sql/map/playtime.sql", map, n)
            .fetch_all(db)
            .await
    }
}
